New Topic: Open Source (#935)
Codecademy/docs: content/open-source/concepts/node-js/node-js.md

---
Title: 'Node.js'
Description: 'Node.js is a free, open-source JavaScript runtime environment for back-end development and building server-side or desktop applications.'
Subjects:
  - 'Web Development'
  - 'Computer Science'
Tags:
  - 'Developer Tools'
  - 'Dependency'
  - 'Node'
  - 'Open Source'
CatalogContent:
  - 'introduction-to-javascript'
  - 'paths/front-end-engineer-career-path'
---

[**Node.js**](https://nodejs.org/en) (or "Node") is a free, open-source [JavaScript (JS)](https://www.codecademy.com/resources/docs/javascript) runtime environment for [back-end](https://www.codecademy.com/resources/docs/general/back-end) development and building [server-side](https://www.codecademy.com/resources/docs/general/server-side-rendering) or desktop applications.

Written in [C](https://www.codecademy.com/resources/docs/c), [C++](https://www.codecademy.com/resources/docs/cpp), and JS, Node.js is built on the open-source [V8](https://v8.dev/) JS engine, used to power browsers such as Google Chrome. As V8 supports new features in JS, they are incorporated into Node.

## JS and Node.js

> JS has existed since 1995 and has since taken over as the dominant language for web development. For much of its life, JS was used mainly for client-side scripting inside [`<script>`](https://www.codecademy.com/resources/docs/html/elements/script) elements executing in web browsers. This limitation meant that developers were often working in many different languages and frameworks between the [front-end](https://www.codecademy.com/resources/docs/general/front-end) (client-side) and back-end (server-side) aspects of a web application.

Author: Looks like this entry already links to General: Back-End towards the top of the entry (on line # 17).

> Although there were other projects to bring JS to server-side applications, the functionality took off with the launch of Node.js in 2009. Node allows developers to write JS code that runs directly in a computer process itself instead of in a browser. Node can, therefore, be used to write server-side applications with access to the operating system, file system, and everything else required to build fully-functional applications.

Author: It looks like much of this entry mirrors [this existing Codecademy blog post about Node.js](https://www.codecademy.com/article/what-is-node).

> Although there were other projects to bring JS to server-side applications, the functionality took off with the launch of Node.js in 2009. Node allows developers to write JS code that runs directly in a computer process itself instead of in a browser. Node can, therefore, be used to write server-side applications with access to the operating system, file system, and everything else required to build fully-functional applications.

Author: I agree, though, that this is worded a little funny. I think I would rephrase as:

"Node allows developers to write server-side JS code that can directly access a computer's operating system and processes such as input/output and file storage."

Node REPL can also load existing JS files. If we had the following code saved into `script.js`:

```js
var a = 'Node REPL is fun!';
```

We can use `.load` to load it into the REPL. `.load` takes a path argument, so to load `script.js` we would use `.load ./script.js`.

```shell
$ node
> .load ./script.js
var a = 'Node REPL is fun!';

> a
'Node REPL is fun!'
```

After the script file is loaded, the variables are accessible in the REPL, so when we evaluate the `a` variable, it's value has been set by loading `script.js`, and 'Node REPL is fun!' prints to the console.

This can be tested by running `node` in a terminal or check out the <a href="https://nodejs.org/api/repl.html" >REPL docs</a> for more functionality.

## Package Management

Node packages are a convenient way to share modules between Node developers. The service [npm](https://www.codecademy.com/resources/docs/javascript/npm) is the default package manager for Node; it becomes available after Node is installed.

npm allows access to the hundreds of thousands of open-source packages available.

In addition to npm, [Yarn](https://yarnpkg.com/en/) is gaining popularity as another JS package manager.

## Node Versions

> Node's major releases aim to support the latest JS features including ES6/ES2015 and beyond without transpilation. A resource called [node.green](http://node.green/) provides a list of supported JS features in different versions of Node.

Author: Can do! I also think we should add a new General concept entry about compilers/transpilers so we can link back to a Docs entry in the future. 😄

### Modules

Node has many built-in modules to aid in interactions with the [terminal](https://www.codecademy.com/resources/docs/general/terminal), the computer file system, and the Internet. These include:

- [HTTP](https://nodejs.org/dist/latest/docs/api/http.html) and [HTTPS](https://nodejs.org/dist/latest/docs/api/https.html) for creating web servers.
- [File System](https://nodejs.org/dist/latest/docs/api/fs.html), [OS](https://nodejs.org/dist/latest/docs/api/os.html), and [Path](https://nodejs.org/dist/latest/docs/api/path.html) for interacting with the file system, operating system, and [file/directory paths](https://www.codecademy.com/resources/docs/general/file-paths).

More information about Node's built-in features can be found [here](https://nodejs.org/dist/latest/docs/api/)

## Why Node?

Per its official home page, Node "uses an event-driven, non-blocking I/O model." In practice, this means that Node is built well to handle asynchronous JS code to perform many asynchronous activities such as reading and writing to the file system, handling connections to database servers, or handling requests as a web server.

> To handle asynchronous code, Node uses a callback-based system. Node functions and methods that will implement some asynchronous activity take a [callback function](https://www.codecademy.com/resources/docs/javascript/callbacks). This function will be called whenever the asynchronous operation has resolved. By convention, the first argument of this callback is an error placeholder. If an error occurred in the asynchronous operation occurred (trying to read a non-existent file, for example), the error argument will be an `Error` object, that will be `null` if no error occurs.

Reviewer: ```suggestion
To handle asynchronous code, Node uses a callback-based system. Node functions and methods that will implement some asynchronous activity take a [callback function](https://www.codecademy.com/resources/docs/javascript/callbacks). This function will be called whenever the asynchronous operation has resolved. By convention, the first argument of this callback is an error placeholder. If an error in the asynchronous operation occurred (trying to read a non-existent file, for example), the error argument will be an `Error` object, that will be `null` if no error occurs.
```

Reviewer: redundant here

For example, if we have the following saved in a `script.js` file:

```js
console.log('I am a Node program');
```

We can run the terminal command in the same folder as `script.js`:

```shell
node script.js
```

This will start Node, print `I am a Node program` to the terminal window, and then exit, as the script file has finished execution.

## Node as a REPL

Node can also be used in a terminal as a read-evaluate-print-loop (REPL). This functionality allows JS commands to be executed from the terminal.

With Node installed, the REPL can be launched with the `node` command in a terminal and pressing <kbd>Enter</kbd>. This launches the interactive JS environment where any valid JS code (e.g., `4 + 5`) can be run. After executing a command, Node will always print the result of that evaluation:

```shell
$ node
> 4 + 5
9
> function nodeIsGreat() {
... console.log('Node is great!');
... }
undefined
> nodeIsGreat()
Node is great!
undefined
> .exit
$
```

In this example, the user launches node on line one with the `node` terminal command.

> On line 2 the user types `4 + 5` and evaluates with the <kbd>Enter</kbd> key. `9` prints to the output terminal.

Reviewer: ```suggestion
On line 2, the user types `4 + 5` and evaluates with the <kbd>Enter</kbd> key. `9` prints to the output terminal.
```

When the function is invoked on line 8, `Node is great!` logs to the console, and `undefined` logs after, because `nodeIsGreat()` returns `undefined`.

To exit the Node REPL, use the `.exit` command at any point and return to the system shell. Pressing <kbd>Ctrl</kbd> + <kbd>C</kbd> twice will also exit.

### Loading Existing Files

Node REPL can also load existing JS files. If we had the following code saved into `script.js`:

```js
var a = 'Node REPL is fun!';
```

We can use `.load` to load it into the REPL. `.load` takes a path argument, so to load `script.js` we would use `.load ./script.js`.

```shell
$ node
> .load ./script.js
var a = 'Node REPL is fun!';

> a
'Node REPL is fun!'
```

After the script file is loaded, the variables are accessible in the REPL, so when we evaluate the `a` variable, it's value has been set by loading `script.js`, and 'Node REPL is fun!' prints to the console.

> This can be tested by running `node` in a terminal or check out the <a href="https://nodejs.org/api/repl.html" >REPL docs</a> for more functionality.

Reviewer: ```suggestion
This can be tested by running `node` in a terminal or check out the <a href="https://nodejs.org/api/repl.html" >REPL docs</a> for more information on commands and functionality.
```

## Package Management

> Node packages are a convenient way to share modules between Node developers. The service [npm](https://www.codecademy.com/resources/docs/javascript/npm) is the default package manager for Node; it becomes available after Node is installed.

Reviewer: ```suggestion
Node packages are a convenient way to share modules between Node developers. The service Node Package Manager ([npm](https://www.codecademy.com/resources/docs/javascript/npm)) is the default package manager for Node; it becomes available after Node is installed.
```

npm allows access to the many open-source packages available.

In addition to npm, [Yarn](https://yarnpkg.com/en/) is gaining popularity as another JS package manager.

## Node Versions

Node's major releases aim to support the latest JS features including ES2015 (ES6) and beyond without transpiling to an earlier JS version at runtime. A resource called [node.green](http://node.green/) provides a list of supported JS features in different versions of Node.

### Version Management

> As with any major release of software, top-level Node releases (8.x, 7.x, etc.) sometimes introduce error-causing changes in applications built in older versions of the environment. A version manager can be used to switch between multiple versions of Node on a single computer.

Reviewer: ```suggestion
As with any major release of software, top-level Node releases (8.x, 7.x, etc.) sometimes introduce error-causing changes in applications built using older versions on an environment. A version manager can be used to switch between multiple versions of Node on a single computer.
```
